source/_posts/2020-07-28-Kubernetes-controllers-A-New-Hope.md

```diff
@@ -0,0 +1,297 @@
+
+{% note %}
+One of WebAssembly's unique attributes is the ability to run sandboxed without relying on OS process boundaries. Requiring a 1-to-1 correspondence between wasm instances and heavyweight OS processes would take away this key advantage for many use cases. Fork/exec are the obvious example of an API that's difficult to implement well if you don't have POSIX-style processes, but a lot of other things in POSIX are tied to processes too. So it isn't a simple matter to take POSIX, or even a simple subset of it, to WebAssembly.
+{% endnote %}
+
+A socket, for example, is a process wide resource. This means that if module A open a TCP socket, module B might access to it if we don't implement proper security countermeasures. 
+
+### Designing the ABI
+
+Given our knowledge and studies of the current state of art for Wasm ABIs, we found out there are 3 possible (non mutually excluded) approaches to the ABI design:
+
+* **High level ABI**: expose the Kubernetes client functionalities
+* **Medium level ABI**: proxy the HTTP client functionalities (send HTTP request, receive HTTP response)
+* **Low level ABI**: proxy the syscalls (`epoll`, `read`, `write`, `bind`, ...)
+
+We think that a good solution should mix all these 3 _"levels"_. Some low-level APIs are always necessary for basic things like logging, get configuration from environment, set timeout and so on. Proxying HTTP can be useful to invoke services outside Kubernetes (e.g. to trigger a cloud vendor API to enable a service).
+Exposing the Kubernetes client APIs unlocks a great potential to optimize the controllers, to name few: host need to watch only one time a `ConfigMap` required by all different controllers, a single cache layer can be shared among all controllers watching a specific namespace, RBAC checks can be done directly in the host all the pods, RBAC checks might be done directly on the host side. All these optimizations might incredibly reduce the network and cpu pressure on Kubernetes clusters.
+
+{% note info %}
+Just as an example, in Knative all the controllers watch a `ConfigMap` called `config-logging` to configure the logging level. Assuming there are 10 Knative controllers in separate processes, all of them needs to do long polling HTTP requests to Kubernetes in order to check for changes and reconfigure logging. With the solution proposed here, we need only one long polling HTTP request, reducing by a factor 10 the network activity and CPU cycles on that particular watch (both server side and client side).
```

Comment:
Knative of course already has that optimization. I'd rather not tie it to the project but rather say "imagine a system consisting of 10 controllers, all watching a configmap to configure loglevels..."